### PR TITLE
provisioner/chef: make the Chef `attributes` param also accept a raw JSON string

### LINKED ISSUE #3074: chef provisioner - attributes from file or template
it would be useful to have an attribute in the chef provisioner to allow setting the chef attributes from a file (like the -j flag you have in chef-client) or to provide the attributes as an string compute from a terraform template resource.

### LINKED ISSUE #3572: Chef Provisioner: Don't overwrite existing /etc/chef/first-boot.json
It would be nice if `terraform-provisioner-chef` did not overwrite an existing `/etc/chef/first-boot.json` when it runs. We would like to dynamically upload a json attributes file based on a var but are unable to because the provisioner replaces the existing file. 

We want to be able to do something like this. 

```
resource "openstack_compute_instance_v2" "server" {
  name = "test-server"
  provisioner "file" {
      source = "chef/${var.chef-attrib-json}"
      destination = "/etc/chef/first-boot.json"
  }
  provisioner "chef" {
    run_list = ["role[test-server]"]
    ...
  }
  ...
}
```

### LINKED ISSUE #4881: Chef Provisioner - certificate verify failed
I'm having some issues with the chef provisioner.  

```
  provisioner "chef" {
    run_list = [""]
    node_name = "${aws_instance.servera.private_dns}"
    secret_key = "${file("~/encrypted_data_bag_secret")}"
    server_url = "chef-serverURL"
    validation_client_name = "terraform-validator"
    validation_key = "${file("./terraform-validator.pem")}"
    ssl_verify_mode = "verify_none"

    connection {
      user = "centos"
      private_key = "${file("./keyfile.pem")}"
      host = "${aws_instance.servera.private_ip}"
      agent = "false"
    }
```

It goes through the chef bootstrap, but fails on the SSL Validation:

Starting Chef Client, version 12.6.0
aws_instance.servera (chef): Creating a new client identity for *removed* using the validator key.
aws_instance.servera (chef): [2016-01-28T00:28:46+00:00] ERROR: SSL Validation failure connecting to host: *removed* - SSL_connect returned=1 errno=0 state=error: certificate verify failed
aws_instance.servera (chef):
aws_instance.servera (chef): ================================================================================
aws_instance.servera (chef): Chef encountered an error attempting to create the client "*removed*"
aws_instance.servera (chef): ================================================================================

aws_instance.servera (chef):
aws_instance.servera (chef): Running handlers:
aws_instance.servera (chef): [2016-01-28T00:28:46+00:00] ERROR: Running exception handlers

Error applying plan:

1 error(s) occurred:

* Command "sudo chef-client -j \"/etc/chef/first-boot.json\" -E \"_default\"" exited with non-zero exit status: 1

I've tried using different validator keys including the one I use on my workstation to bootstrap nodes with knife.  I've set ssl_verify_mode to verify_none, and tried changing permissions on my validator key file.  Also tried with:

validation_key = "${file("terraform-validator.pem")}"
validation_key = "terraform-validator.pem"

 I am not sure what I configured incorrectly.

### PR DESCRIPTION
**Pull request by svanharmelen**
See the updated docs for more details and examples, but in short this enables the `attributes` param from the Chef provisioner to accept a raw JSON string.

Fixes #3074
Fixes #3572
Fixes #4881

### CODE REVIEW

**Comment by svanharmelen** [comment]
@phinze whenever you have time, could you double check this one? Thanks!

**Comment by svanharmelen** [comment]
@phinze could you check this one if you have a minute? If OK we can close some issue :wink: Thx!

**Comment by svanharmelen** [comment]
@hashicorp/terraform-committers would anyone that can spare a minute have a look at this PR before I merge it? Would be nice to have someone do a small review, so we go move ahead and close a couple of related issues.

Thanks!

**Comment by apparentlymart** [comment]
After a quick pass it seems pretty reasonable to me.

My only feedback is rather subjective: should we make the new JSON-string version of this attribute be called `attributes_json` for consistency with the usage in the Chef *provider*? The rationale for that naming was to make it clear to someone reading config that valid JSON is expected here, rather than requiring them to read the docs to figure that out. But regardless of the original merits of that naming decision, it might be nice to take this opportunity to be consistent.

This might make the deprecation easier to explain in the docs and error message too: then we can just say that `attributes` is deprecated and that `attributes_json` should be used instead, whereas the docs in your patch don't mention the *old* attributes mechanism at all; that could be confusing for someone who finds themselves maintaining an older config and sees that the config and the docs disagree but gets no explanation of why.

Not rigid on this, so feel free to ignore me and proceed if you disagree. :grinning:

**Comment by svanharmelen** [comment]
@apparentlymart thanks for the review!

Personally I'm not to fond of the name `attributes_json`, but your arguments seem to make sense :wink: Will let it marinate for tonight (well after office hours here) and (most likely) update the PR tomorrow...

Thx!

**Comment by phinze** [comment]
Finally managed to swing around to this! +1 on `attributes_json` for same reasons @apparentlymart brings up. Otherwise LGTM.

**Comment by svanharmelen** [comment]
@apparentlymart @phinze please have one more look to confirm this is ok now...

**Comment by svanharmelen** [comment]
Anyone have a second verify the changes? Would really like to close this one and cleanup the related issues...

File: builtin/provisioners/chef/resource_provisioner.go [hunk 1]
```diff
@@ -222,6 +236,10 @@ func (r *ResourceProvisioner) Validate(c *terraform.ResourceConfig) (ws []string
 		ws = append(ws, "secret_key_path is deprecated, please use "+
 			"secret_key instead and load the key contents via file()")
 	}
+	if _, ok := c.Config["attributes"]; ok {
+		ws = append(ws, "using map style attribute values is deprecated, "+
+			" please use a single raw JSON string instead")
```
Comment (apparentlymart) [inline builtin/provisioners/chef/resource_provisioner.go]: I guess this should now say "please use attributes_json instead"

**Comment by apparentlymart** [comment]
This seems reasonable to me! (minor error message comment notwithstanding)

I notice that the certificate-related rider on this patch introduces a new dependency on `knife`. Is it guaranteed that a chef client distribution always includes knife, and so it'll get installed automatically as part of installing the rest of it? (I don't have any chef-managed systems handy right now to check, so please forgive my ignorance!)

**Comment by svanharmelen** [comment]
@apparentlymart yes, `knife` is part of `chef-client`, so it's save to assume the availability after `chef-client` is installed :wink:

See [here](https://docs.chef.io/chef_client.html#knife-title-subcommands) for more details, mainly:

```
The chef-client includes two knife commands for managing SSL certificates:

Use knife ssl check to troubleshoot SSL certificate issues
Use knife ssl fetch to pull down a certificate from the Chef server to the /.chef/trusted_certs directory on the workstation.

```

**Pull request merged**